refactor(decorations): move provider code

lewis6991 opened the pull request:
Move decoration provider code to a separate file.

lewis6991 committed:
refactor(decorations): move provider code

Move decoration provider code to a separate file.

bfredl commented on src/nvim/decoration/providers.c:
```diff
@@ -0,0 +1,227 @@
+// This is an open source non-commercial project. Dear PVS-Studio, please check
```
subdir is overkill, perhaps `src/nvim/decoration_provider.c` ?

bfredl commented on src/nvim/screen.c:
```diff
@@ -2231,34 +2151,11 @@ static int win_line(win_T *wp, linenr_T lnum, int startrow, int endrow, bool noc
-      decor_add_ephemeral(lnum-1, 0, lnum-1, 0, &err_decor);
+      provider_err_virt_text(lnum, provider_err);
       provider_err = NULL;
-      has_decor = true;
```
I think this is still necessary.

lewis6991 replied:
I've incorporated this into `decor_provider_invoke_line`

lewis6991 committed:
refactor(decorations): unglobal decor_providers

Now all the decoration provider code is in one place, decor_providers no
longer need to be global.
